internal/graph/cached_resolver_test.go

```diff
@@ -0,0 +1,514 @@
+					},
+				},
+			},
+			subsequentReq: &ResolveCheckRequest{
+				StoreID:              "12",
+				AuthorizationModelID: "33",
+				TupleKey:             tuple.NewTupleKey("document:abc", "reader", "user:XYZ"),
+				ContextualTuples: []*openfgav1.TupleKey{
+					{
+						Object:   "document:aaa",
+						Relation: "reader",
+						User:     "user:XYZ",
+					},
+					{
+						Object:   "document:xxx",
+						Relation: "reader",
+						User:     "user:XYZ",
+					},
+					{
+						Object:   "document:yyy",
+						Relation: "reader",
+						User:     "user:XYZ",
+					},
+				},
+			},
+			setInitialResult: func(mock *MockCheckResolver, request *ResolveCheckRequest) {
+				mock.EXPECT().ResolveCheck(ctx, request).Times(1).Return(result, nil)
+			},
+			setTestExpectations: func(mock *MockCheckResolver, request *ResolveCheckRequest) {
+				mock.EXPECT().ResolveCheck(ctx, request).Times(1).Return(result, nil)
+			},
+		},
+		{
+			name: "first_contextual_tuples_then_no_contextual_tuples_does_not_return_results_from_cache",
+			initialReq: &ResolveCheckRequest{
+				StoreID:              "12",
+				AuthorizationModelID: "33",
+				TupleKey:             tuple.NewTupleKey("document:abc", "reader", "user:XYZ"),
+				ContextualTuples: []*openfgav1.TupleKey{
+					{
+						Object:   "document:aaa",
+						Relation: "reader",
+						User:     "user:XYZ",
+					},
+					{
+						Object:   "document:xxx",
+						Relation: "reader",
+						User:     "user:XYZ",
+					},
+				},
+			},
+			subsequentReq: &ResolveCheckRequest{
+				StoreID:              "12",
+				AuthorizationModelID: "33",
+				TupleKey:             tuple.NewTupleKey("document:abc", "reader", "user:XYZ"),
+				ContextualTuples:     []*openfgav1.TupleKey{},
+			},
+			setInitialResult: func(mock *MockCheckResolver, request *ResolveCheckRequest) {
+				mock.EXPECT().ResolveCheck(ctx, request).Times(1).Return(result, nil)
+			},
+			setTestExpectations: func(mock *MockCheckResolver, request *ResolveCheckRequest) {
+				mock.EXPECT().ResolveCheck(ctx, request).Times(1).Return(result, nil)
+			},
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+
+			initialMockResolver := NewMockCheckResolver(ctrl)
+			initialReq := req
+			if test.initialReq != nil {
+				initialReq = test.initialReq
+			}
+			test.setInitialResult(initialMockResolver, initialReq)
+
+			// expect first call to result in actual resolve call
+			dut := NewCachedCheckResolver(initialMockResolver,
+				WithLogger(logger.NewNoopLogger()),
+				WithMaxCacheSize(10))
+			defer dut.Close()
+
+			_, _ = dut.ResolveCheck(ctx, initialReq)
+
+			newCtrl := gomock.NewController(t)
+			defer newCtrl.Finish()
+
+			newResolver := NewMockCheckResolver(newCtrl)
+			test.setTestExpectations(newResolver, test.subsequentReq)
+
+			dut2 := NewCachedCheckResolver(newResolver, WithExistingCache(dut.cache))
```

Comment:
Should we be testing this with the same pattern we use in `pkg/server/server.go` and in `pkg/server/commands/list_objects.go`?

Namely,

```go
localChecker := NewLocalCheckResolver(...)
cachedChecker := NewCachedCheckResolver(localChecker, ...)
localChecker.SetDelegate(cachedChecker)

_, _ = cachedChecker.ResolveCheck(ctx, test.initialReq)
...
result, err := cachedChecker.ResolveCheck(ctx, test.subsequentReq)
require.Nil(t, err)
require.Equal(...)
```

Reply:
@jon-whit : I am not sure what you are asking here - the purpose of this test is to assert the number of times ResolveCheck is called in order to validate whether data is served from cache vs. from DS.  As such, we need to mock the check resolver.

Reply:
Sorry I wasn’t clear. What I meant was to try mimic the pattern we use more closely, which implies not creating a second mock CheckResolver. It’s not necessary.

Instead, you can
```go
mockCheckResolver := NewMockCheckResolver(ctrl)
initialReq := req
if test.initialReq != nil {
	initialReq = test.initialReq
}
test.setInitialResult(mockCheckResolver, initialReq)

// expect first call to result in actual resolve call
dut := NewCachedCheckResolver(mockCheckResolver,
	WithLogger(logger.NewNoopLogger()),
	WithMaxCacheSize(10))
defer dut.Close()

_, _ = dut.ResolveCheck(ctx, initialReq)

test.setTestExpectations(mockCheckResolver, test.subsequentReq)

dut2 := NewCachedCheckResolver(dut, WithExistingCache(dut.cache))
defer dut2.Close()

actualResult, err := dut2.ResolveCheck(ctx, test.subsequentReq)
require.Equal(t, result.Allowed, actualResult.Allowed)
require.Nil(t, err)
```

e.g.
Use `dut` as the delegate for the `dut2` CheckResolver. That way we’re composing/chaining the resolvers in a way that exercises the delegate behavior.